libkatana_python_native/include/katana/python/NumbaSupport.h

```diff
@@ -0,0 +1,145 @@
+  template <typename Return, typename... Extra>
+  void def_func(
+      pybind11::module_& m, const char* name, Return (*f)(Args...),
+      const Extra&... extra) const {
+    m.def(name, f, extra...);
+    auto func = m.attr(name);
+    auto numba_support =
+        pybind11::module_::import("katana.native_interfacing.numba_support");
+    numba_support.attr("register_function")(
+        func, (uintptr_t)f, 0,
+        PythonTypeTraits<remove_cvref_t<Return>>::ctypes_type(),
+        PythonTypeTraits<remove_cvref_t<Args>>::ctypes_type()...);
+  }
+
+public:
+  // TODO(amp): The generated wrappers used from numba code are created per
+  //  *signature*, not per function. So two functions will the same overall
+  //  signature (including containing class) will use the same wrapper. This
+  //  could possibly cause a megamorphic call site that defeats the CPU branch
+  //  predictor. The fix for this is to somehow distinguish different
+  //  functions/methods at the template level. This could be done with a counter
+  //  or a name argument of some kind.
+
+  template <typename Return, typename... Extra>
+  void operator()(
+      pybind11::module_& m, const char* name, Return (*pf)(Args...),
+      const Extra&... extra) const {
+    def_func(m, name, pf, extra...);
+  }
+
+  template <typename Func, typename... Extra>
+  void operator()(
+      pybind11::module_& m, const char* name, Func&& f,
+      const Extra&... extra) const {
+    this->template operator()(m, name, &f, extra...);
+  }
+
+  template <typename Cls, typename Return, typename ImplCls, typename... Extra>
+  void operator()(
+      pybind11::class_<Cls>& cls, const char* name,
+      Return (ImplCls::*pmf)(Args...), const Extra&... extra) const {
+    // This leaks a single pointer sized struct for each defined numba function.
+    // Repeated import could theoretically cause this to matter, but it's very
+    // unlikely.
+    def_class_method<Cls, Return>(
+        cls, name, pmf, new MemberFunction<ImplCls, Return, Args...>{pmf},
+        extra...);
+  }
+
+  template <typename Cls, typename Return, typename ImplCls, typename... Extra>
+  void operator()(
+      pybind11::class_<Cls>& cls, const char* name,
+      Return (ImplCls::*pmf)(Args...) const, std::true_type,
+      const Extra&... extra) const {
+    // As above.
+    def_class_method<Cls, Return>(
+        cls, name, pmf, new ConstMemberFunction<ImplCls, Return, Args...>{pmf},
+        extra...);
+  }
+};
+
+/// Declare a method or function to be called from Numba and Python.
+///
+/// This should be called the same way `pybind11`'s `def` function is called.
+///
+/// \tparam Args The argument types of the function.
+template <typename... Args>
+constexpr DefWithNumbaImpl<Args...> DefWithNumba{};
+
+/// Register a Python class for use from Numba compiled code.
+///
+/// This calls `katana.native_interfacing.numba_support.register_class`
+template <typename T>
+void
+RegisterNumbaClass(pybind11::class_<T>& cls) {
```

Comment:
What exactly does this function do?

Reply:
It sets up a class so that `DefWithNumba` can be used on it's methods. This should be documented explicitly.